neat_cache/lib/src/providers/redis.dart

```diff
@@ -0,0 +1,154 @@
+
+  Future<_RedisContext> _context;
+  bool _isClosed = false;
+
+  RedisCacheProvider(
+    this._connectionString, {
+    Duration connectTimeLimit = const Duration(seconds: 30),
+    Duration commandTimeLimit = const Duration(milliseconds: 200),
+    Duration reconnectDelay = const Duration(seconds: 30),
+  })  : _connectTimeLimit = connectTimeLimit,
+        _commandTimeLimit = commandTimeLimit,
+        _reconnectDelay = reconnectDelay {
+    assert(_connectionString.isNotEmpty, 'connectionString must be given');
+    assert(!_connectTimeLimit.isNegative, 'connectTimeLimit is negative');
+    assert(!_commandTimeLimit.isNegative, 'commandTimeLimit is negative');
+    assert(!_reconnectDelay.isNegative, 'reconnectDelay is negative');
+  }
+
+  Future<_RedisContext> _createContext() async {
+    try {
+      print('### CONNECTING!!!');
+      final connection = await Connection.connect(_connectionString)
+          .timeout(_connectTimeLimit);
+      // Create context
+      final client = Client(connection);
+      return _RedisContext(
+        connection: connection,
+        client: client,
+        commands: client.asCommands(),
+      );
+    } on TimeoutException {
+      throw IntermittentCacheException('connect failed with timeout');
+    } on SocketException {
+      throw IntermittentCacheException('connect failed with socket exception');
+    } on Exception {
+      throw IntermittentCacheException('connect failed with exception');
+    }
+  }
+
+  Future<_RedisContext> _getContext() {
+    if (_context != null) {
+      return _context;
+    }
+    _context = _createContext();
+    scheduleMicrotask(() async {
+      _RedisContext ctx;
+      try {
+        ctx = await _context;
+      } on IntermittentCacheException {
+        // If connecting fails, then we sleep and try again
+        await Future.delayed(_reconnectDelay);
+        _context = null; // reset _context, so next operation creates a new
+        return;
+      } catch (e) {
+        _logger.shout('unknown error/exception connecting to redis', e);
+        _context = null; // reset _context, so next operation creates a new
+        rethrow; // propagate the error to crash to application.
+      }
+      // If connecting was successful, then we await the connection being
+      // closed or error, and we reset _context.
+      // Any errors have already been returned if any commands were affect
+      // otherwise the error is just a broken TCP connection and we ignore it.
+      try {
+        await ctx.connection.done;
+      } catch (e) {
+        // ignore error
+      }
+      _context = null;
+    });
+    return _context;
+  }
+
+  Future<T> _withRedis<T>(_Op<T> fn) async {
+    if (_isClosed) {
+      throw StateError('CacheProvider.closed() has been called');
+    }
+    final ctx = await _getContext();
+    try {
+      return await fn(ctx.commands).timeout(_commandTimeLimit);
+    } on TimeoutException {
+      print('############ OP TIMED OUT');
```

Comment:
Use logger instead.

Reply:
These were 100% debug statements, for me when I was fixing the debugging the test... I thought I had removed them :see_no_evil:  :see_no_evil:  :see_no_evil:  :see_no_evil: :see_no_evil: